pkg/reports/store.go

```diff
@@ -52,39 +60,58 @@ func (s *Store) Write(ctx context.Context, workload kube.Object, reports vulnera
 		reportName := fmt.Sprintf("%s-%s-%s", strings.ToLower(string(workload.Kind)),
 			workload.Name, containerName)
 
-		vulnerabilityReport := &starboardv1alpha1.VulnerabilityReport{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      reportName,
-				Namespace: workload.Namespace,
-				Labels: labels.Set{
-					kube.LabelResourceKind:      string(workload.Kind),
-					kube.LabelResourceName:      workload.Name,
-					kube.LabelResourceNamespace: workload.Namespace,
-					kube.LabelContainerName:     containerName,
+		vulnerabilityReport := &starboardv1alpha1.VulnerabilityReport{}
+
+		err := s.client.Get(ctx, types.NamespacedName{Name: reportName, Namespace: workload.Namespace}, vulnerabilityReport)
+		if errors.IsNotFound(err) {
+			vulnerabilityReport = &starboardv1alpha1.VulnerabilityReport{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      reportName,
+					Namespace: workload.Namespace,
+					Labels: labels.Set{
+						kube.LabelResourceKind:      string(workload.Kind),
+						kube.LabelResourceName:      workload.Name,
+						kube.LabelResourceNamespace: workload.Namespace,
+						kube.LabelContainerName:     containerName,
+						etc.LabelPodSpecHash:        hash,
+					},
 				},
-			},
-			Report: report,
-		}
-		err = controllerutil.SetOwnerReference(owner, vulnerabilityReport, s.scheme)
-		if err != nil {
-			return err
+				Report: report,
+			}
+			err = controllerutil.SetOwnerReference(owner, vulnerabilityReport, s.scheme)
+			if err != nil {
+				return err
+			}
+			log.Info("Creating VulnerabilityReport",
+				"report", fmt.Sprintf("%s/%s", workload.Namespace, reportName),
+				"hash", hash)
+			err := s.client.Create(ctx, vulnerabilityReport)
+			if err != nil {
+				return err
+			}
+			return nil
 		}
 
-		err := s.client.Create(ctx, vulnerabilityReport)
-		if err != nil {
-			return err
-		}
+		// Do not modify the object that might be cached.
```

Comment:
Is there a possibility that the set of vulnerabilities has changed, even if the pod spec hasn't?

Reply:
I believe it's not possible in our implementation. However, this is just a good practice to clone objects returned from cache to avoid subtle bugs that are hard to debug.